README.md
# GitHub Action - Update VersionInfo Files
This GitHub action updates version info in source files based on the provided semantic version. This can be used before building to include the build information in released binaries.

![Github JavaScript Actions CI/CD](https://github.com/dolittle/update-version-info-action/workflows/Github%20JavaScript%20Actions%20CI/CD/badge.svg)

### Pre requisites
Create a workflow `.yml` file in your `.github/workflows` directory. An [example workflow](#example-workflow) is available below.

For more information, reference the GitHub Help Documentation for [Creating a workflow file](https://help.github.com/en/articles/configuring-a-workflow#creating-a-workflow-file)

### Inputs
- `version` (required): Version of the release to use to replace values in the specified files.
- `files-to-update` (required): A comma or newline separated list of files to replace values in.
- `allow-multiple-replacements`: Whether or not to allow multiple replacements of each value in a file. Defaults to 'false'.
- `allow-no-replacements`: Whether or not to allow no replacements of each value in a file. Defaults to 'false'.
- `replacements`: A comma separated list of values to replace. Items can be any of (major, minor, patch, prerelease, full). Defaults to 'major, minor, patch, prerelease'.
- `major-type`: The type of literal to replace for the 'major' replacement. Can be (string, number, exact). Defaults to 'number'.
- `major-match`: The value to find for the 'major' replacement. Defaults to '377'.
- `minor-type`: The type of literal to replace for the 'minor' replacement. Can be (string, number, exact). Defaults to 'number'.
- `minor-match`: The value to find for the 'minor' replacement. Defaults to '389'.
- `patch-type`: The type of literal to replace for the 'patch' replacement. Can be (string, number, exact). Defaults to 'number'.
- `patch-match`: The value to find for the 'patch' replacement. Defaults to '368'.
- `prerelease-type`: The type of literal to replace for the 'prerelease' replacement. Can be (string, number, exact). Defaults to 'string'.
- `prerelease-match`: The value to find for the 'prerelease' replacement. Defaults to 'PRERELEASE'.
- `full-type`: The type of literal to replace for the 'full' replacement. Can be (string, number, exact). Defaults to 'string'.
- `full-match`: The value to find for the 'full' replacement. Defaults to '377.389.368-PRERELEASE'.

### Outputs
This action has no outputs.

### Example Workflow
```yaml
on:
  push:
    branches:
    - '**'
  pull_request:
    types: [closed]

name: GitHub action workflow name

jobs:
  context:
    name: Job name
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v2
      - name: Establish context
        id: context
        uses: dolittle/establish-context-action@v2
      - name: Increment version
        id: increment-version
        uses: dolittle/increment-version-action@v2
        with:
          version: ${{ steps.context.outputs.current-version }}
          release-type: ${{ steps.context.outputs.release-type }}
      - name: Update versioninfo
        uses: dolittle/update-version-info-action@v1
        with:
          version: ${{ steps.increment-version.outputs.next-version }}
          files-to-update: |
            Source/VersionInfo.cs
            Source/VersionInfo.go
            Source/VersionInfo.ts
            Source/VersionInfo.js
        
```
## Contributing
We're always open for contributions and bug fixes!

### Pre requisites
node <= 12
yarn
git


CHANGELOG.md
# [1.0.2] - 2022-12-7 [PR: #3](https://github.com/dolittle/update-version-info-action/pull/3)
## Summary

Update deps and modernise


# [1.0.1] - 2021-12-1 [PR: #2](https://github.com/dolittle/update-version-info-action/pull/2)
## Summary

The `semver` library exposes the `prerelease` of the parsed versions as an array of elements. `.toString()` joined these with a comma making the replaced prerelease value strange. This PR fixes that by using `.join('.')` instead.

### Fixed

- The replaced `prerelease` strings now contains `.` instead of `,`.


# [1.0.0] - 2021-12-1 [PR: #1](https://github.com/dolittle/update-version-info-action/pull/1)
## Summary

Introducing the "Update VersionInfo Files" action. This GitHub action can be used to replace values in source code from the current build pipeline to include build information in published binaries.

See the [README](README.md) for a description on how to use it.




Source/Inputs.ts
// Copyright (c) Dolittle. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import { getBooleanInput, getInput } from '@actions/core';
import semver, { SemVer } from 'semver';
import { ILogger } from '@dolittle/github-actions.shared.logging';
import { ReplacementConfig } from './Configuration/ReplacementConfig';
import { isReplacement, Replacement } from './Configuration/Replacement';
import { isReplacementType, ReplacementType } from './Configuration/ReplacementType';

/**
 * Represents the inputs to the action.
 */
export class Inputs {
    /**
     * Initialises a new instance of the {@link Inputs} class.
     * @param {SemVer} version - The version to update the version info files with.
     * @param {string[]} filesToUpdate - The version info files to update.
     * @param {ReplacementConfig[]} replacements - The replacements to perform.
     * @param {boolean} allowMultipleReplacements - Whether or not to allow multiple replacements to occur in a file.
     * @param {boolean} allowNoReplacements - Whether or not to allow no replacements to occur in a file.
     */
    constructor(
        readonly version: SemVer,
        readonly filesToUpdate: string[],
        readonly replacements: ReplacementConfig[],
        readonly allowMultipleReplacements: boolean,
        readonly allowNoReplacements: boolean
    ) {}

    /**
     * Logs the values of the inputs to the provided logger.
     * @param {ILogger} logger - The {@link ILogger} to write to.
     */
    log(logger: ILogger): void {
        logger.info('Inputs:');
        logger.info(`  version: '${this.version}'`);
        logger.info(`  files-to-update: '${this.filesToUpdate.join(', ')}'`);
        logger.info('  replacements:');
        this.replacements.forEach((replacement => {
            logger.info(`    ${replacement.replacement} will replace ${replacement.match} using ${replacement.type}`);
        }));
        logger.info(`  allow-multiple-replacements: ${this.allowMultipleReplacements}`);
        logger.info(`  allow-no-replacements: ${this.allowNoReplacements}`);
    }

    /**
     * Parses the inputs provided to the action.
     * @returns {Inputs} - The parsed {@link Inputs}.
     */
    static parse(): Inputs {
        return new Inputs(
            this.getVersion(),
            this.getFilesToUpdate(),
            this.getReplacements(),
            this.getOptionalBooleanInput('allow-multiple-replacements'),
            this.getOptionalBooleanInput('allow-no-replacements')
        );
    }

    private static getVersion(): SemVer {
        const input = getInput('version', { required: true });

        if (!semver.valid(input)) {
            throw new Error(`The provided version '${input}' is not a valid SemVer.`);
        }

        return semver.parse(input)!;
    }

    private static getFilesToUpdate(): string[] {
        const input = getInput('files-to-update', { required: true });

        return this.splitTrimIgnoreEmpty(input, /[,\n]/);
    }

    private static getReplacements(): ReplacementConfig[] {
        const input = getInput('replacements', { required: false });

        const configs: ReplacementConfig[] = [];
        for (const replacement of this.splitTrimIgnoreEmpty(input, ',')) {
            if (isReplacement(replacement)) {
                const config = this.getReplacementConfig(replacement);
                configs.push(config);
            } else {
                throw new Error(`The provided replacement ${replacement} is not a valid replacement`);
            }
        }

        if (configs.length > 0) {
            return configs;
        }

        return [
            this.getReplacementConfig('major'),
            this.getReplacementConfig('minor'),
            this.getReplacementConfig('patch'),
            this.getReplacementConfig('prerelease'),
        ];
    }

    private static getReplacementConfig(replacement: Replacement): ReplacementConfig {
        const typeInput = getInput(replacement + '-type', { required: false }).trim();

        if (typeInput !== '' && !isReplacementType(typeInput)) {
            throw new Error(`The provided replacement type ${typeInput} for ${replacement} is not a valid replacement type`);
        }

        const type = typeInput !== '' ? typeInput : this.getDefaultReplacementType(replacement);

        const matchInput = getInput(replacement + '-match', { required: false }).trim();

        const match = matchInput !== '' ? matchInput : this.getDefaultReplacementMatch(replacement);

        return { replacement, type, match };
    }

    private static getDefaultReplacementType(replacement: Replacement): ReplacementType {
        switch (replacement) {
            case 'major':
            case 'minor':
            case 'patch':
                return 'number';
            case 'prerelease':
            case 'full':
                return 'string';
        }
    }

    private static getDefaultReplacementMatch(replacement: Replacement): string {
        switch (replacement) {
            case 'major':
                return '377';
            case 'minor':
                return '389';
            case 'patch':
                return '368';
            case 'prerelease':
                return 'PRERELEASE';
            case 'full':
                return '377.389.368-PRERELEASE';
        }
    }

    private static splitTrimIgnoreEmpty(input: string, separator: string | RegExp): string[] {
        const split = input.split(separator);
        const trimmed = split.map(_ => _.trim());
        const filtered = trimmed.filter(_ => _.length > 0);
        return filtered;
    }

    private static getOptionalBooleanInput(name: string): boolean {
        if (getInput(name, { required: false }) === '') {
            return false;
        }

        return getBooleanInput(name);
    }
}
